internal/herd_regression_test.ml

```diff
@@ -32,66 +34,100 @@ type flags = {
 
 (* Utilities. *)
 
-let litmuses_of_dir dir =
-  let all_files = Array.to_list (Sys.readdir dir) in
-  let only_litmus = List.filter TestHerd.is_litmus all_files in
-  let full_paths = List.map (Filename.concat dir) only_litmus in
-  List.sort String.compare full_paths
+let for_each_litmus_in_dir dir f =
+  let handle = Unix.opendir dir in
+  let rec for_each_litmus' () =
+    try
+      let name = Unix.readdir handle in
+      begin if TestHerd.is_litmus name then
+        f (Filename.concat dir name)
+      end ;
+      for_each_litmus' ()
+    with
+    | End_of_file -> ()
```

Comment:
The  above code is not tail-recursive, due to the `try ... with ...` the result of the call to `forach_litnus'` is not a call to `foreach_litmus'`. Hence, it still may suffer from stack overflow. Here is a tail-recursive version:
```OCaml

let readdir d = try Some (Unix.readdir d) with End_of_file -> None

let for_each d f =
  let rec for_each () =
     match readdir d with
      | Some name -> if TesHerd.is_litmus then begin f name end ;  foreach ()
      | None -> () in
  for_each ()
 
```

Reply:
Done. Thanks!